iree/hal/cuda/graph_command_buffer.c

```diff
@@ -0,0 +1,382 @@
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_cuda_graph_command_buffer_update_buffer(
+    iree_hal_command_buffer_t* base_command_buffer, const void* source_buffer,
+    iree_host_size_t source_offset, iree_hal_buffer_t* target_buffer,
+    iree_device_size_t target_offset, iree_device_size_t length) {
+  return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                          "need cuda implementation");
+}
+
+static iree_status_t iree_hal_cuda_graph_command_buffer_copy_buffer(
+    iree_hal_command_buffer_t* base_command_buffer,
+    iree_hal_buffer_t* source_buffer, iree_device_size_t source_offset,
+    iree_hal_buffer_t* target_buffer, iree_device_size_t target_offset,
+    iree_device_size_t length) {
+  iree_hal_cuda_graph_command_buffer_t* command_buffer =
+      iree_hal_cuda_graph_command_buffer_cast(base_command_buffer);
+
+  CUdeviceptr target_device_buffer = iree_hal_cuda_buffer_device_pointer(
+      iree_hal_buffer_allocated_buffer(target_buffer));
+  target_offset += iree_hal_buffer_byte_offset(target_buffer);
+  CUdeviceptr source_device_buffer = iree_hal_cuda_buffer_device_pointer(
+      iree_hal_buffer_allocated_buffer(source_buffer));
+  source_offset += iree_hal_buffer_byte_offset(source_buffer);
+  CUDA_MEMCPY3D params = {};
+  params.Depth = 1;
+  params.Height = 1;
+  params.WidthInBytes = length;
+  params.dstDevice = target_device_buffer;
+  params.srcDevice = source_device_buffer;
+  params.srcXInBytes = source_offset;
+  params.dstXInBytes = target_offset;
+  params.srcMemoryType = CU_MEMORYTYPE_DEVICE;
+  params.dstMemoryType = CU_MEMORYTYPE_DEVICE;
+  // Serialize all the nodes for now.
+  CUgraphNode dep[] = {command_buffer->lastNode};
+  size_t numNode = command_buffer->lastNode ? 1 : 0;
+  CUDA_RETURN_IF_ERROR(
+      command_buffer->context->syms,
+      cuGraphAddMemcpyNode(&command_buffer->lastNode, command_buffer->graph,
+                           dep, numNode, &params,
+                           command_buffer->context->cu_context),
+      "cuGraphAddMemcpyNode");
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_cuda_graph_command_buffer_push_constants(
+    iree_hal_command_buffer_t* base_command_buffer,
+    iree_hal_executable_layout_t* executable_layout, iree_host_size_t offset,
+    const void* values, iree_host_size_t values_length) {
+  return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                          "need cuda implementation");
+}
+
+static iree_status_t iree_hal_cuda_graph_command_buffer_push_descriptor_set(
+    iree_hal_command_buffer_t* base_command_buffer,
+    iree_hal_executable_layout_t* executable_layout, uint32_t set,
+    iree_host_size_t binding_count,
+    const iree_hal_descriptor_set_binding_t* bindings) {
+  iree_hal_cuda_graph_command_buffer_t* command_buffer =
+      iree_hal_cuda_graph_command_buffer_cast(base_command_buffer);
+  void** arguments;
+
+  iree_status_t status = iree_allocator_malloc(
```

Comment:
each push_descriptor_set will leak the previous allocation like this

push_descriptor_set is incremental: you could have push([bindings 0-2]) + push([bindings 3-4]) - so you'll want to allocate this once at command buffer creation and then just set the array elements in push/bind, vs allocating anything new/replacing existing sets

the easiest thing is to pick a max (64, etc) and then just embed that in the struct - then no need for allocations.

Reply:
Good catch, I'm allocation it at command buffer time now.